impl/src/test/java/org/ehcache/integration/SerializerCountingTest.java

```diff
@@ -0,0 +1,195 @@
+    );
+
+    cache.put(42L, "TheAnswer");
+    printSerializationCounters("Put Offheap");
+    cache.get(42L);
+    printSerializationCounters("Get Offheap fault");
+    cache.get(42L);
+    printSerializationCounters("Get Offheap faulted");
+
+    cache.put(42L, "Wrong ...");
+    printSerializationCounters("Put OffHeap (update faulted)");
+  }
+
+  @Test
+  public void testSerializerOffHeapOnHeapCopyPutGet() {
+    Cache<Long, String> cache = cacheManager.createCache("offHeap", newCacheConfigurationBuilder()
+            .withResourcePools(newResourcePoolsBuilder().heap(10, EntryUnit.ENTRIES).offheap(10, MemoryUnit.MB))
+            .add(new DefaultCopierConfiguration(SerializingCopier.class, CopierConfiguration.Type.KEY))
+            .add(new DefaultCopierConfiguration(SerializingCopier.class, CopierConfiguration.Type.VALUE))
+            .buildConfig(Long.class, String.class)
+    );
+
+    cache.put(42L, "TheAnswer");
+    printSerializationCounters("Put Offheap");
+    cache.get(42L);
+    printSerializationCounters("Get Offheap fault");
+    cache.get(42L);
+    printSerializationCounters("Get Offheap faulted");
+
+    cache.put(42L, "Wrong ...");
+    printSerializationCounters("Put OffHeap (update faulted)");
+  }
+
+  private void printSerializationCounters(String operation) {
+    System.out.println("Operation " + operation);
+    System.out.println(format("Key Serialization - %d / Deserialization - %d / Equals - %d", CountingSerializer.keySerializeCounter.get(), CountingSerializer.keyDeserializeCounter.get(), CountingSerializer.keyEqualsCounter.get()));
+    System.out.println(format("Value Serialization - %d / Deserialization - %d / Equals - %d", CountingSerializer.serializeCounter.get(), CountingSerializer.deserializeCounter.get(), CountingSerializer.equalsCounter.get()));
+    CountingSerializer.serializeCounter.set(0);
+    CountingSerializer.deserializeCounter.set(0);
+    CountingSerializer.equalsCounter.set(0);
+    CountingSerializer.keySerializeCounter.set(0);
+    CountingSerializer.keyDeserializeCounter.set(0);
+    CountingSerializer.keyEqualsCounter.set(0);
+  }
+
+  @Serializer.Transient
+  @Serializer.Persistent
+  public static class CountingSerializer<T> implements Serializer<T> {
+
+    static AtomicInteger serializeCounter = new AtomicInteger(0);
+    static AtomicInteger deserializeCounter = new AtomicInteger(0);
+    static AtomicInteger equalsCounter = new AtomicInteger(0);
+    static AtomicInteger keySerializeCounter = new AtomicInteger(0);
+    static AtomicInteger keyDeserializeCounter = new AtomicInteger(0);
+    static AtomicInteger keyEqualsCounter = new AtomicInteger(0);
+
+    private final JavaSerializer<T> serializer;
+
+    public CountingSerializer(ClassLoader classLoader) {
+      serializer = new JavaSerializer<T>(classLoader);
+    }
+
+    @Override
+    public ByteBuffer serialize(T object) throws SerializerException {
+      if (PRINT_STACK_TRACES) {
```

Comment:
Replacing the ```System.out``` calls with assertions is probably a good addition for a later time, when we're confident we can't do any better. For now, I'd leave this test as-is.